### PR TITLE
test: Post review comments for uncovered code additions

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

File: test/run [hunk 1]
```diff
@@ -16,8 +16,8 @@ case $TEST_SCENARIO in
         [ "${TEST_SCENARIO}" = "${TEST_SCENARIO%%devel}" ] || export NODE_ENV=development
         [ "${TEST_SCENARIO}" = "${TEST_SCENARIO%%devel}" ] || OPTS="$OPTS --coverage"
         [ "${TEST_SCENARIO}" = "${TEST_SCENARIO##firefox}" ] || export TEST_BROWSER=firefox
-        test/image-prepare --verbose $TEST_OS
-        test/common/run-tests --jobs ${TEST_JOBS:-1} --test-dir test/verify ${OPTS}
+        test/image-prepare --verbose --quick $TEST_OS
```
Comment (mvollmer) [inline test/run:19]: Hi

**Review by cockpituous — CHANGES_REQUESTED** [review]
Not good enough!

**Review by cockpituous — CHANGES_REQUESTED** [review]
Not good enough!

File: test/common/run-tests [hunk 2]
```diff
@@ -21,6 +21,8 @@ parent.ensure_bots()  # NOQA: testvm lives in bots/
 import testlib
 import testvm
 
+from task import github
+
```
Comment (cockpituous) [inline test/common/run-tests:25]: Uncovered

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/nav.jsx [hunk 3]
```diff
@@ -48,6 +48,7 @@ export class CockpitNav extends React.Component {
         // Click on active menu item (when using arrows to navigate through menu)
         function clickActiveItem() {
             const cur = document.activeElement;
+            console.log("This wont run");
```
Comment (cockpituous) [inline pkg/shell/nav.jsx:51]: Uncovered

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/nav.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:51]: Uncovered

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/nav.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:51]: Uncovered

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/nav.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:51]: This added line is not executed by any test.

**Review by cockpituous — CHANGES_REQUESTED** [review]

**Review by cockpituous — CHANGES_REQUESTED** [review]

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/nav.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:51]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220627-124556-94ed2deb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (cockpituous) [inline pkg/shell/nav.jsx:51]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220627-130821-94ed2deb-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by cockpituous — CHANGES_REQUESTED** [review]

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/hosts.jsx [hunk 4]
```diff
@@ -127,6 +127,7 @@ export class CockpitHosts extends React.Component {
     filterHosts(host, term) {
         if (!term)
             return host;
+        console.log("Another unexecuted line");
```
Comment (cockpituous) [inline pkg/shell/hosts.jsx:130]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-061558-24f3995a-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:51]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-061558-24f3995a-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/hosts.jsx [hunk 5]
```diff
@@ -127,6 +127,7 @@ export class CockpitHosts extends React.Component {
     filterHosts(host, term) {
         if (!term)
             return host;
+        console.log("A added line");
```
Comment (cockpituous) [inline pkg/shell/hosts.jsx:130]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-064306-0d71ea51-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx [hunk 6]
```diff
@@ -30,6 +30,11 @@ export const SidebarToggle = () => {
     );
 };
 
+function uncalled_function() {
+    console.log("This whole function should not be called");
+    console.log("And another line to make a bigger code block");
+}
```
Comment (cockpituous) [inline pkg/shell/nav.jsx:36]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-064306-0d71ea51-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx [hunk 7]
```diff
@@ -48,6 +53,9 @@ export class CockpitNav extends React.Component {
         // Click on active menu item (when using arrows to navigate through menu)
         function clickActiveItem() {
             const cur = document.activeElement;
+            console.log("A added line, it might be executed");
+            if (cur === true)
+                uncalled_function();
```
Comment (cockpituous) [inline pkg/shell/nav.jsx:58]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-064306-0d71ea51-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/nav.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:36]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-065219-4598d09e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/hosts.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/shell/hosts.jsx:130]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-083430-8f8981f7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:36]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-083430-8f8981f7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx (hunk 7 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:58]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-083430-8f8981f7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by mvollmer** [comment]
Hmm, we should probably update the links in old comments, instead of just skipping new comments for the same locations. Or just delete all old comments and make new ones.

File: pkg/shell/hosts.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/shell/hosts.jsx:130]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-161940-d45b541e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:36]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-161940-d45b541e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx (hunk 7 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:58]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-161940-d45b541e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by cockpituous — CHANGES_REQUESTED** [review]

File: pkg/shell/nav.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:36]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-171014-d45b541e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/shell/nav.jsx (hunk 7 above)
Comment (cockpituous) [inline pkg/shell/nav.jsx:58]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17475-20220628-171014-d45b541e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by mvollmer** [comment]
I have tested this a bunch with the changes introduced by the `DBG` commit here, and it all seems to work as expected. The bots will delete their old comments before making new ones, and that doesn't seem to leave any annoying traces.
I think the next step would be to install this in `main` and try it out in a couple of real pull requests. if anything bad happens we can revert this quickly, or hopefully adjust it.

**Comment by martinpitt** [comment]
@mvollmer Agreed -- this is really cool! Please drop the debugging leftover and the DBG commit, drop no-test, rebase, and I'll happily approve. As you said it's easy to re-disable if it wreaks havoc, but let's test this in the field.

File: test/common/lcov.py [hunk 8]
```diff
@@ -358,10 +371,98 @@ def covranges(functions):
...
+            subprocess.check_call(["genhtml", "--no-function-coverage",
+                                   "--prefix", os.getcwd(),
+                                   "--title", title,
+                                   "--output-dir", f"{output}/Coverage", all_file])
+            comments = get_review_comments(diff_file)
+            rev = os.environ.get("TEST_REVISION", None)
+            pull = os.environ.get("TEST_PULL", None)
+            if rev and pull:
+                api = github.GitHub()
+                old_comments = api.get(f"pulls/{pull}/comments") or []
```
Comment (jelly) [inline test/common/lcov.py:458]: From the docs, I've found out, this is limited to 30 comments, so one has to fetch more pages using `page` or setting `per_page` to 100.

https://docs.github.com/en/rest/pulls/comments#list-review-comments-on-a-pull-request

File: test/common/lcov.py [hunk 9]
```diff
@@ -358,10 +371,98 @@ def covranges(functions):
...
+                for oc in old_comments:
+                    if ("body" in oc and "path" in oc and "line" in oc and
+                        "not executed by any test." in oc["body"]):
+                        api.delete(f"pulls/comments/{oc['id']}")
```
Comment (jelly) [inline test/common/lcov.py:462]: So this basically deletes old cockpituous comments right? Wouldn't it be easier to filter on `user.login` I think that should be `cockpituous`.

Comment (mvollmer) [reply test/common/lcov.py:462]: Yeah, I was thinking the same.... will it always be "cockpituous"? And we shouldn't delete all cockpituous comments, only the code coverage ones, so we still would need to check the body, no?

File: test/common/lcov.py [hunk 10]
```diff
@@ -17,21 +17,29 @@
...
 
+import parent
+from task import github
+
+BASE_DIR = os.path.realpath(f'{__file__}/../../..')
+
 __all__ = (
-    "write_lcov"
+    "write_lcov",
+    "get_review_comments"
```
[the remark this answers is not in the document]
Comment (jelly) [reply test/common/lcov.py]: Turns out it's only for star imports and doesn't help with `normal` imports.

File: test/common/lcov.py (hunk 8 above)
Comment (jelly) [reply test/common/lcov.py]: 100 is probably good enough for now yes :+1:

**Review by jelly — APPROVED** [review]

**Comment by mvollmer** [comment]
Hmm, the cockpituous reviews are still blocking the merging of this PR.  Let's not use CHANGES_REQUESTED for them, but just COMMENT.  Then, I'll do some testing again.  To get rid of the bogus reviews that block this, I'll do that on a new PR.

**Comment by mvollmer** [comment]
> To get rid of the bogus reviews that block this, I'll do that on a new PR.

#17523

**Pull request closed without merging**